lib/diaspora_federation/salmon/xml_payload.rb

```diff
@@ -85,19 +93,42 @@ def self.entity_class_name(term)
       # @param [Nokogiri::XML::Element] node xml nodes
       # @return [Entity] instance
       def self.populate_entity(klass, node)
-        data = {}
-        klass.class_props.each do |prop_def|
-          name = prop_def[:name]
+        # Build an auxillary hash to check whether XML node is a defined member of klass
+        props = klass.class_props.map { |prop_def|
           type = prop_def[:type]
-
           if type == String
-            data[name] = parse_string_from_node(prop_def[:xml_name], node)
+            [prop_def[:xml_name], prop_def]
           elsif type.instance_of?(Array)
-            data[name] = parse_array_from_node(type, node)
+            [type.first.entity_name.to_sym, prop_def]
```

Comment:
Maybe we can move the `Array` and `Entity` logic here to `PropertiesDSL` and set the `xml_name` in `define_property`. Then we can always use `prop_def[:xml_name]` here. That would simplify this part. But this is just an idea, I'm not sure if this is the best solution, but I think the `populate_entity` method is too complex.

Another idea is to add an `xml_props` method to `PropertiesDSL` ...

But I'm tired now, maybe I have a better idea tomorrow. Let me know what you think about it.